src/common/atomic_queue.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * atomic_queue.h
+ * Copyright (c) 2020 Hewlett Packard Enterprise Development, LP. All rights
+ * reserved. Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * See https://spdx.org/licenses/BSD-3-Clause
+ *
+ */
+
+#ifndef ATOMIC_QUEUE_H_
+#define ATOMIC_QUEUE_H_
+#include <atomic>
+#include <fam/fam_exception.h>
+#include <fam/fam.h>
+#include <allocator/memserver_allocator.h>
+#include "fam_ops_libfabric.h"
+
+#define MAX_NODE_ADDR_SIZE 64 //#define FT_MAX_CTRL_MSG 64
+#define MAX_DATA_IN_MSG 128
+#define ANONYMOUS_UNION_SIZE 32
+#define ATOMIC_REGION "ATOMIC_REGION"
+#define ATOMIC_REGION_ID 17
+#define ATOMIC_REGION_SIZE 1073741824
+#define MAX_ATOMIC_THREADS 32
+#define QSIZE 20
+#define ATOMIC_WRITE_IN_PROGRESS 64
+#define ATOMIC_WRITE_COMPLETED 128
+#define ATOMIC_BUFFER_ALLOCATED 256
+
+#define ATOMIC_READ 1
+#define ATOMIC_WRITE 2
+#define ATOMIC_SCATTER_STRIDE 4
+#define ATOMIC_SCATTER_INDEX 8
+#define ATOMIC_GATHER_STRIDE 16
+#define ATOMIC_GATHER_INDEX 32
+
+#define ATOMIC_CONTAIN_DATA 256
+
+#define ATOMIC_QUEUE_EMPTY 10
+#define DATAITEMNOTFOUND 11
+#define BUFFERALLOCATEERROR 12
+using namespace std;
+namespace openfam {
+
+typedef struct atomicMsg {
+    short int flag; // bit 0: read/write; 1:data in yes/no; bit 2: inprogress(1)/completed(0)
```

Comment:
you can describe the flag in the macro/enum definition place. So that you can use the type to define the flag and one can read the description of the type at its definition place.